src/extension/app-schema/app-schema-web/src/main/java/org/geoserver/complex/ComplexToSimpleFeatureCollection.java

```diff
@@ -0,0 +1,216 @@
+        this.featureType = requireNonNull(buildConvertedType());
+        LOG.fine(() -> "Converted feature type: " + featureType);
+    }
+
+    /**
+     * Builds and returns the simple feature type based on the complex type and the transformation
+     * rules.
+     */
+    protected SimpleFeatureType buildConvertedType() {
+        FeatureTypeConverter converter =
+                new FeatureTypeConverter(delegate.getSchema(), rulesMap, namespaceSupport);
+        return converter.produceSimpleType();
+    }
+
+    @Override
+    public SimpleFeatureType getSchema() {
+        return featureType;
+    }
+
+    @Override
+    public SimpleFeatureIterator features() {
+        return new ComplexToSimpleFeatureIterator(delegate.features());
+    }
+
+    @Override
+    public SimpleFeatureCollection subCollection(Filter filter) {
+        return null;
+    }
+
+    @Override
+    public SimpleFeatureCollection sort(SortBy order) {
+        return null;
+    }
+
+    @Override
+    public String getID() {
+        return delegate.getID();
+    }
+
+    @Override
+    public void accepts(FeatureVisitor visitor, ProgressListener progress) throws IOException {
+        DataUtilities.visit(this, visitor, progress);
+    }
+
+    @Override
+    public ReferencedEnvelope getBounds() {
+        return delegate.getBounds();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return delegate.contains(o);
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> o) {
+        return delegate.containsAll(o);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public int size() {
+        return delegate.size();
+    }
+
+    @Override
+    public Object[] toArray() {
+        return getFeaturesList().toArray();
+    }
+
+    @Override
+    public <O> O[] toArray(O[] a) {
+        return getFeaturesList().toArray(a);
+    }
+
+    private List<SimpleFeature> getFeaturesList() {
+        try (SimpleFeatureIterator featureIterator = features()) {
+            List<SimpleFeature> featuresList = new ArrayList<>();
+            while (featureIterator.hasNext()) {
+                featuresList.add(featureIterator.next());
+            }
+            return featuresList;
+        }
+    }
+
+    /** Feature iterator wrapper, converts every complex feature into a simple feature. */
+    class ComplexToSimpleFeatureIterator implements SimpleFeatureIterator {
+
+        private FeatureIterator<Feature> delegate;
+
+        public ComplexToSimpleFeatureIterator(FeatureIterator<Feature> delegate) {
+            this.delegate = delegate;
+        }
+
+        public boolean hasNext() {
+            return delegate.hasNext();
+        }
+
+        public SimpleFeature next() throws NoSuchElementException {
+            SimpleFeature simpleFeature = convert(delegate.next());
+            LOG.finer(() -> "Converted simple feature: " + simpleFeature);
```

Comment:
Careful here, it's true that the string concatenation does not happen unless needed, but the creation of the lambda ([which is, itself, also an object](https://www.tutorialspoint.com/are-lambda-expressions-objects-in-java#:~:text=Yes%2C%20any%20lambda%20expression%20is,it%20like%20any%20other%20object.), instantiated and carried around) is not avoided. For something happening for all returned features, it should be guarded.